File: Arrays/ValidParentheses.java
// 20. Valid Parentheses

// Time Complexity  O(n)
// Space Complexity O(n)

class Solution {
    public boolean isValid(String s) {
        if (s.length() % 2 != 0) {
            return false;
        }
        
        HashMap<String, String> map = new HashMap<>();
        Stack<String> stack = new Stack<>();
        map.put("{", "}");
        map.put("(", ")");
        map.put("[", "]");

        for (int i = 0; i < s.length(); i++) {
            String currentParentheses =  String.valueOf(s.charAt(i));
            if (map.get(currentParentheses) != null) {
                stack.push(currentParentheses);
            } else {
                if (stack.empty()) {
                    return false;
                }
                String lastElement = stack.peek();
                if (currentParentheses.equals(map.get(lastElement))) {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }

        return stack.empty();
    }
}

File: Arrays/ValidParentheses.js
// 20. Valid Parentheses

// Time Complexity  O(n)
// Space Complexity O(n)

var isValid = function (s) {
    if (s.length % 2 !== 0) {
        return false;
    }

    const map = {
        "{": "}",
        "[": "]",
        "(": ")",
    };
    const stack = [];

    for (let i = 0; i < s.length; i++) {
        if (map[s[i]]) {
            stack.push(s[i]);
        } else {
            const lastElement = stack[stack.length - 1];
            if (s[i] === map[lastElement]) {
                stack.pop();
            } else {
                return false;
            }
        }
    }

    return stack.length === 0;
};
